Add proposal for low-level PMP configuration API

baiqiang728 opened the pull request:
Two APIs are added for this feature:

* csi_mpu_config_region is used to configure PMP range attribute
* csi_mpu_disable_region is used to disable PMP range by the index of PMP

baiqiang728 committed:
Add proposal for low-level PMP configuration API

Two APIs are added for this feature:

* csi_mpu_config_region is used to configure PMP range attribute
* csi_mpu_disable_region is used to disable PMP range by the index of
PMP

baiqiang728 committed:
Add proposal for low-level PMP configuration API

Two APIs are added for this feature:

* csi_mpu_config_region is used to configure PMP range attribute
* csi_mpu_disable_region is used to disable PMP range by the index of
PMP

WiseCrohn commented on api/C/include/csi_ll_pmp.h:
```diff
@@ -0,0 +1,66 @@
+ * @param enable: Eanble flag to indicate whether this ranged can be accessed or not
+ * @return : Status code
+ */
+csi_status_t csi_mpu_config_region(unsigned int idx, unsigned long base_addr,
```
This API is OK for applications with a fixed PMP setup.  But it is slow, so maybe not suitable for systems where the PMP setup changes when context-switching between tasks.  Perhaps there should be another function with similar parameters, which can be used to construct a PMP setup structure in memory; so the data in this structure can then be written quickly into the CSRs?

VincentCui632 replied:
For such application in MCU, it is better to realize functions by assemble code instead of LL code.  For PMP context size, it is related with CPU PMP configuration.  I strongly recommend PMP entry size is 16 at first, it is popular configuration in practise else.

WiseCrohn replied:
I don't agree that assembly code is necessary to deal with this, it can all be more easily and efficiently handled from C code.  You can allow the BSP to control context size - simply let the BSP define a context structure which the user application instantiates and passes in to the API for initialization.  But most uses of the PMP require dynamic switching of PMP setup according to what process is running.  This means constructing the register values in advance, to make context switch more efficient.
If we add checking for invalid setup into the definition of what this function does, then this makes it even more inefficient to execute on a context switch.

WiseCrohn replied:
To be clear, I am suggesting 2 additional C APIs: one with a similar parameter set to this one, which can be used to construct a PMP setup within a context structure.  And another which quickly and efficiently writes the contents of that context structure into the CSRs.

fanghuaqi replied:
I agree with @WiseCrohn to provide some lower level pmp api and a little higher level api for flexiable configuration of pmp region configuration.

In NMSIS, we defined it like these 

1. PMP API Header file: https://github.com/Nuclei-Software/NMSIS/blob/master/NMSIS/Core/Include/core_feature_pmp.h#L58-L379
2. PMP API Doc: https://doc.nucleisys.com/nmsis/core/api/core_pmp.html
3. PMP Example: https://github.com/Nuclei-Software/nuclei-sdk/blob/master/application/baremetal/demo_pmp/demo_pmp.c#L53-L129

WiseCrohn commented:
There are many ways in which the user could try to create an invalid setup (overlapping regions, etc.).  Does this API attempt to trap all such invalid setups and return with an error?

baiqiang728 commented:
> There are many ways in which the user could try to create an invalid setup (overlapping regions, etc.). Does this API attempt to trap all such invalid setups and return with an error?

Yes I thinks so. Users who implements these APIs should be able to check the parameters whether it is invalidate or not(such as  overlapping regions, index out of range).

WiseCrohn commented:
Would further APIs similar to freedom metal APIs be useful? (https://sifive.github.io/freedom-metal-docs/apiref/pmp.html)

As a minimum I think an API to discover the number of PMP regions is useful.  Also, a function to change a region's attributes.

baiqiang728 committed:
Add proposal for low-level PMP configuration API

These API can be classified to two types, users can choose different API
at different stage:

* set or get PMP configurations by one API. It can be used as system
initialization stage.
* The others are used for setting or getting specific parameter of PMP
configuratons. These may be can be used at context switch and make it
more effectively

baiqiang728 commented:
> Would further APIs similar to freedom metal APIs be useful? (https://sifive.github.io/freedom-metal-docs/apiref/pmp.html)
> 
> As a minimum I think an API to discover the number of PMP regions is useful. Also, a function to change a region's attributes.
Thanks @WiseCrohn ， I've update a new commit. And there are two types of APIs: 
1. set or get PMP configurations by one API. It can be used as system initialization stage.
2. The others are used for setting or getting specific parameter of PMP configuratons. These may be can be used at context switch and make it more effectively

VincentCui632 commented on api/C/include/csi_ll_pmp.h:
```diff
@@ -0,0 +1,66 @@
+ * @param idx: The index of PMP config entries
+ * @return : Status code
+ */
+csi_status_t csi_mpu_disable_region(unsigned int idx);
```
Is it better to replace "mpu” with "pmp"  ?

WiseCrohn commented:
> And there are two types of APIs:
> set or get PMP configurations by one API. It can be used as system initialization stage.
> The others are used for setting or getting specific parameter of PMP configuratons. These may be can be used at
> context switch and make it more effectively

I don't see this commit, maybe you didn't push it yet?  But I don't think what you describe here is what is needed.

A context switch isn't just about moving or changing one region.  It is a switch between 2 different complete PMP configurations, each associated with the execution context of a running process.  Any firmware managing this process will, I believe, want to maintain multiple shadow register sets corresponding to the different contexts.

What's needed therefore is a function to help with constructing one of these shadow register sets in memory, and another to help with loading it into the PMP registers.

baiqiang728 committed:
Add proposal for low-level PMP configuration API

Two type of API is added:

1. Set or get all of the content of PMP range.
2. Set or get a specific content of PMP range. Such as address mode ,
readable etc.

VincentCui632 commented on api/C/include/csi_ll_pmp.h:
```diff
@@ -34,33 +34,204 @@ typedef struct {
+ * @return : Status code
+ */
+csi_status_t csi_pmp_get_region(unsigned int idx, unsigned long *base_addr,
+                                   unsigned long *size, mpu_region_attr_t *attr);
```
it should be "pmp_region_attr_t" instead of "mpu_region_attr_t"

VincentCui632 commented on api/C/include/csi_ll_pmp.h:
```diff
@@ -34,33 +34,204 @@ typedef struct {
+ * @return : Status code
+ */
+csi_status_t csi_pmp_config_region(unsigned int idx, unsigned long base_addr,
+                                   unsigned long size, mpu_region_attr_t *attr);
```
it should be "pmp_region_attr_t" instead of "mpu_region_attr_t"

VincentCui632 commented on api/C/include/csi_ll_pmp.h:
```diff
@@ -34,33 +34,204 @@ typedef struct {
-csi_status_t csi_mpu_config_region(unsigned int idx, unsigned long base_addr,
-                                   unsigned long size, mpu_region_attr_t *attr,
-                                   unsigned int enable);
+csi_status_t csi_pmp_set_writeable(unsigned int idx, bool flag_w);
```
writable

baiqiang728 replied:
Hi Vincent，

I've fixed them. Pls help to check them agian

Thanks

VincentCui632 commented on api/C/include/csi_ll_pmp.h:
```diff
@@ -34,33 +34,204 @@ typedef struct {
  * @return : Status code
  */
-csi_status_t csi_mpu_config_region(unsigned int idx, unsigned long base_addr,
-                                   unsigned long size, mpu_region_attr_t *attr,
```
it should be "pmp_region_attr_t" instead of "mpu_region_attr_t"

VincentCui632 commented on api/C/include/csi_ll_pmp.h:
```diff
@@ -34,33 +34,204 @@ typedef struct {
  * @return : Status code
  */
-csi_status_t csi_mpu_disable_region(unsigned int idx);
+csi_status_t csi_pmp_get_writeable(unsigned int idx, bool *flag_w);
```
writable

baiqiang728 committed:
Modify the name of strcut and API name

* "mpu" in struct is changed to "pmp"
* "writeable" in API is changed to "writable"

VincentCui632 reviewed (commented):
All looks good to me. but I wonder to know if it needs adding static inline attribution for some function to improve performance in implemenation.

WiseCrohn commented:
> All looks good to me. but I wonder to know if it needs adding static inline attribution for some function to improve performance in implemenation.

In my opinion these functions are not good candidates for static inline: see, for example, the discussion here: https://stackoverflow.com/questions/28926828/when-to-use-static-inline-instead-of-regular-functions .  It is generally best to let the compiler optimize your code as it sees fit, unless you have good reason to think that you know better.

In any case these functions are not performant.  As I said above, for a performant system you need to construct the register values "offline" in memory then load them as complete set.  I want to support this use case.  If you don't want to add it in this PR, then I will propose it as a separate PR.

WiseCrohn commented on api/C/include/csi_ll_pmp.h:
```diff
@@ -0,0 +1,237 @@
+} csi_pmp_region_attr_t;
+
+/*
+ * Get PMP range number for the hard id.
```
What is "range number"?  Does this just means how many regions there are?

baiqiang728 replied:
Yes， it means how many ranges there are

cubidesj commented on api/C/include/csi_ll_pmp.h:
```diff
@@ -0,0 +1,237 @@
+/*
+ * Get PMP range number for the hard id.
+ *
+ * If the hard id is invalidate, such as out of support range, it should return
```
I guess this is "hart id". I guess that this matches the contents of mhartid, right?

baiqiang728 replied:
Thanks, I fixed it now

fanghuaqi commented on api/C/include/csi_ll_pmp.h:
```diff
@@ -0,0 +1,237 @@
+ * Attribution of PMP coniguration
+ */
+typedef struct {
+    bool flag_r; /* The flag of readable. 0: not readable; 1: readable */
```
Here I prefer to list these flags_r/w/x/lock into one variable which match the PMP configuration register format, so user can easily set the pmpcfg using this flags value without complex if else calculation.

1. ![image](https://github.com/riscv-non-isa/riscv-rvm-csi/assets/1538922/c1bbb4b8-435e-461c-a97f-ff55feb4593b)
2. The generated encoding.h also contains pmp attribute macros which can be easily used https://github.com/riscv/riscv-opcodes/blob/master/encoding.h#L276-L285

VincentCui632 replied:
@baiqiang728 Using bit flag to operate with such PMP flag is best choice.

WiseCrohn replied:
I agree - then have one function to set the configuration instead of lots of little functions dealing with each flag.

WiseCrohn replied:
For the moment though, define our own enumeration rather than trying to include encodings.h.  We will auto-generate a similar file to encodings.h but this will happen later.

baiqiang728 replied:
fixed.
I've changed to one function to set the configuration instead of lots of little functions dealing with each flag

fanghuaqi commented on api/C/include/csi_ll_pmp.h:
```diff
@@ -0,0 +1,237 @@
+ * @param range_number: A pointer to store the ouptut value of PMP range number
+ * @return : Status code
+ */
+csi_status_t csi_pmp_get_range_num(unsigned int hart_id, unsigned int *range_number);
```
what do you mean by range number, does it mean the pmp entries supported? So this function is expected to do csr pmp register prope to get this entries number?

baiqiang728 replied:
Yes， it means the pmp entries supported. I think it could be implemented by doing csr pmp probe like opensbi or just a platform define like `#define PMP_COUNT			64`

WiseCrohn replied:
I think it would be better if this function returns the value instead of dereferencing a pointer.  Then the compiler can optimize it out, if the implementation in the BSP is simply a `#define`.

baiqiang728 replied:
> I think it would be better if this function returns the value instead of dereferencing a pointer. Then the compiler can optimize it out, if the implementation in the BSP is simply a `#define`.
Hi @WiseCrohn ，
My purpose to use error code is to avoid the case that invalid hart_id is input. So if I change this function to returns the value instead of dereferencing a pointer ,  is that suitable that it returns zero when invalid hart_id is input?

cubidesj replied:
I had written something along those lines replying to @WiseCrohn, I agree that it is simpler to just return but that implies that we have to decide and document what happens if the value for an invalid input (or other error) would be. I don't have a strong preference either way but, whenever possible we should try to be coherent in our approach.

WiseCrohn replied:
I think return an integer, and have -1 as an error code (e.g. invalid hart ID), and leave 0 to represent the case where there is no PMP (so there are no regions!).  This is analogous to what we have done elsewhere, e.g. csi_get_irq_priority etc.

baiqiang728 replied:
fixed. I've changed this API.  It wil return the value of PMP entry of specified hart id, and will return -1 if the input is invalid

fanghuaqi commented on api/C/include/csi_ll_pmp.h:
```diff
@@ -0,0 +1,237 @@
+ * Configure PMP entry.
+ *
+ * This function only can be called in M-mode. If it is called
+ * in other mode, it may return an error or cause carsh. When this function is called
```
I think only one result when called in other mode should be defined, such as it is not permit call it in other mode except m, it will just crash.

fanghuaqi commented on api/C/include/csi_ll_pmp.h:
```diff
@@ -0,0 +1,237 @@
+ *
+ * This function only can be called in M-mode. If it is called
+ * in other mode, it may return an error or cause carsh. When this function is called
+ * with success, it will not be effective untill the core has been exit M-mode.
```
the core has been exit M-mode -> the core exits M-mode

fanghuaqi commented on api/C/include/csi_ll_pmp.h:
```diff
@@ -0,0 +1,237 @@
+    bool flag_r; /* The flag of readable. 0: not readable; 1: readable */
+    bool flag_w; /* The flag of writable. 0: not writable; 1: writable */
+    bool flag_x; /* The flag of execable. 0: not execable; 1: execable */
+    csi_addr_matching_t mode; /* Address matching mode */
```
This mode should be decided when you configure pmp, you can judge it by the address you pass.

![image](https://github.com/riscv-non-isa/riscv-rvm-csi/assets/1538922/e1994e53-6844-4db3-ac7e-46ce6c7069ec)

fanghuaqi commented on api/C/include/csi_ll_pmp.h:
```diff
@@ -0,0 +1,237 @@
+ * @param attr: A point to the attribution of this entry to be configured
+ * @return : Status code
+ */
+csi_status_t csi_pmp_config_region(unsigned int idx, unsigned long base_addr,
```
prefer to use csi_pmp_set_region, since below api are set/get like

ghost commented:
> candidates

fanghuaqi commented on api/C/include/csi_ll_pmp.h:
```diff
@@ -0,0 +1,237 @@
+ * @param mode: The address mode of PMP config entries
+ * @return : Status code
+ */
+csi_status_t csi_pmp_set_address_mode(unsigned int idx, csi_addr_matching_t mode);
```
The mode is bind with the address, so it should be set together with the address, and be automatically calculated. see https://github.com/riscv-software-src/opensbi/blob/master/lib/sbi/riscv_asm.c#L328

baiqiang728 replied:
@fanghuaqi ， I thought about it before. But there is another mode TOR. If there is no address mode input,  the driver does not know whether it should be set to TOR or other mode.  Do you have any suggestion about it ? Or we just add a separate API using two PMP entries to do that as you said

fanghuaqi commented on api/C/include/csi_ll_pmp.h:
```diff
@@ -0,0 +1,237 @@
+ *              this entry
+ * @return : Status code
+ */
+csi_status_t csi_pmp_get_address_mode(unsigned int idx, csi_addr_matching_t *mode);
```
No useful, since we have csi_pmp_get_region

fanghuaqi commented on api/C/include/csi_ll_pmp.h:
```diff
@@ -0,0 +1,237 @@
+ * @param flag_x: The excecutable flag of PMP config entries.
+ * @return : Status code
+ */
+csi_status_t csi_pmp_set_executable(unsigned int idx, bool flag_x);
```
`csi_pmp_set/get_executable/writable/readable/lock` maybe can be set as one such as csi_pmp_set/get to set the pmp attributes.

fanghuaqi reviewed (changes requested):
And we also missed an API to set TOR address mode, since it need two address passed and use two pmp entry.

cubidesj commented on api/C/include/csi_ll_pmp.h:
```diff
@@ -0,0 +1,237 @@
+csi_status_t csi_pmp_get_range_num(unsigned int hart_id, unsigned int *range_number);
+
+/*
+ * Configure PMP entry.
```
A silly question about vocabulary: What is the relationship between "entry", "range" and "region"? I have seen the three used in the API and I want to be sure that we have a common understanding on what those words means and what they imply.

baiqiang728 replied:
I change all "range" to "entry" as it is specified in RISC-V Privileged Architecture

cubidesj commented on api/C/include/csi_ll_pmp.h:
```diff
@@ -0,0 +1,237 @@
+#define CSI_LL_PMP_H
+
+#include "csi_types.h"
+#include <stdint.h>
```
It is a silly question but why are we including this if we are not using the types that are defined in there?

WiseCrohn replied:
csi_status_t comes from there

baiqiang728 committed:
Change the PMP API
a. csi_pmp_get_entry_num is used to get the PMP entry number of
specified hart id.
b. csi_pmp_set_entry is used to set the PMP configuration information of
specified PMP index
c. csi_pmp_get_entry is used to get the PMP configuration information of
specified PMP index

WiseCrohn commented:
@baiqiang728 the way this is intended to work is that the parser processes the YAML to produce both the C code and the asciidoc (pdf).  Currently, you are not actually auto-generating either the C-code or the asciidoc.  You seem to have created the C-code by hand.

To use the auto-generation you need to add csi_ll_pmp-spec.yaml to the list of modules at the bottom of rvm-csi-spec.yaml.  Then run make.  You will find there are some build errors.  This is because your c-type-declarations doesn't conform to the schema (no name field).  To solve this I think it would be better to use an enum type declaration for these.  This would group things together in the desired way.

When you have it building, you also need to add auto-gen/modules/csi_ll_pmp_h.adoc[] to the list of included modules at the bottom of chapter2.adoc.  This then gives rise to further problems because your attempt to create tables in the function comments does not render correctly in asciidoc.  If you want to include tables like this then we need to think about how that should be done.

baiqiang728 committed:
Fixed the compiler error of csi_ll_pmp-spec.yaml
a.csi_ll_pmp-spec.yaml is added to rvm-csi-spec.yaml to auto-generate
csi_ll_pmp.h
a.csi_ll_pmp_h.adoc[] is added to chapter2.adoc

baiqiang728 commented:
> @baiqiang728 the way this is intended to work is that the parser processes the YAML to produce both the C code and the asciidoc (pdf). Currently, you are not actually auto-generating either the C-code or the asciidoc. You seem to have created the C-code by hand.
> 
> To use the auto-generation you need to add csi_ll_pmp-spec.yaml to the list of modules at the bottom of rvm-csi-spec.yaml. Then run make. You will find there are some build errors. This is because your c-type-declarations doesn't conform to the schema (no name field). To solve this I think it would be better to use an enum type declaration for these. This would group things together in the desired way.
> 
> When you have it building, you also need to add auto-gen/modules/csi_ll_pmp_h.adoc[] to the list of included modules at the bottom of chapter2.adoc. This then gives rise to further problems because your attempt to create tables in the function comments does not render correctly in asciidoc. If you want to include tables like this then we need to think about how that should be done.

Thanks @WiseCrohn  for your detail explanation. I've submit a new patch to solve this two issues just now. Would you pls help me to check it again ?

WiseCrohn commented on api/C/include/csi_ll_pmp.h:
```diff
@@ -0,0 +1,88 @@
+ * @return : The number of PMP entry; -1 will be returned as and error code if the
+ * input is invalid.
+ */
+int csi_pmp_get_entry_num(unsigned int hart_id);
```
Please change the name of this function to csi_pmp_get_num_entries, and change the text to say "number of PMP entries" rather than "range number" or "entry number", this is clearer.

WiseCrohn commented on api/C/include/csi_ll_pmp.h:
```diff
@@ -0,0 +1,88 @@
+
+/*
+ * Set PMP entry. This function only can be called in M-mode. If it is called in
+ * other mode, it may cause carsh. When this function is called with success, it
```
Rather than saying that accessing in wrong mode will cause a crash, would it be more accurate to say it will cause an access fault exception?  This doesn't necessarily imply a crash.

WiseCrohn commented on api/C/include/csi_ll_pmp.h:
```diff
@@ -0,0 +1,88 @@
+ * PMP_ATTR_FLAG_X: indicate that the PMP entry permits to execute PMP_ATTR_FLAG_L:
+ * indicate that the PMP entry permits to lock
+ */
+#define PMP_ATTR_FLAG_R				    _UL(0x01)
```
Dos this _UL() mean cast to unsigned long?  If so, why?  I think these values are intended to OR together to form an unsigned char to pass into pmp_attr.

WiseCrohn commented on api/C/include/csi_ll_pmp.h:
```diff
@@ -0,0 +1,88 @@
+ * PMP_ATTR_ADDR_MATCHING_NAPOT: indicate that the PMP entry's address-matching is
+ * NAPOT
+ */
+#define PMP_ATTR_ADDR_MATCHING_TOR		_UL(0x08)
```
Should add PMP_ATTR_ADDR_MATCHING_DISABLED here as value 0.

WiseCrohn commented on api/C/include/csi_ll_pmp.h:
```diff
@@ -0,0 +1,88 @@
+ * or size is not supported or over overlapped etc.
+ *
+ * @param idx: The PMP entry index
+ * @param base_addr: The base address of this range need to be configured
```
If the address matching is TOR, what does it do with base address?  Is it ignored, or does it modify the address of the previous entry?

baiqiang728 replied:
Yes, my idea is that it should be ignored. And the address of previous entry is considered as base address. Do you have good idea about it?

WiseCrohn replied:
I think this is probably OK if the behaviour is explained in the comments.

WiseCrohn commented on api/C/include/csi_ll_pmp.h:
```diff
@@ -0,0 +1,88 @@
+ * other mode, it may cause carsh. When this function is called with success, it
+ * will not be effective untill  the core exits M-mode. Also, an error should be
+ * returned if the output parameters is invalid, such as idx out of range, address
+ * or size is not supported or over overlapped etc.
```
Not clear what "over overlapped" means?  In hardware, ranges are allowed to overlap, and lower-indexed entries are prioritized in this case.

baiqiang728 committed:
Change the comments of APIs

WiseCrohn commented:
Please merge latest changes from main branch and resolve the conflict.

baiqiang728 committed:
Merge branch 'riscv-non-isa:main' into main

baiqiang728 committed:
Merge branch 'main' into pmp

WiseCrohn commented:
Please reinstate the launch.json that has been deleted on this branch, this file is useful for debugging the python parser, and is not doing any harm.

baiqiang728 committed:
Add launch.json

baiqiang728 commented:
> Please reinstate the launch.json that has been deleted on this branch, this file is useful for debugging the python parser, and is not doing any harm.

Hi @WiseCrohn ，

I added it back just now .

fanghuaqi commented on api/csi_ll_pmp-spec.yaml:
```diff
@@ -0,0 +1,127 @@
+      If the hart id is invalid, such as out of support range, it should return
+      an error
+    c-params:
+    - name: hart_id
```
Why `hart_id` is required, if you running on one hart, you can only access to the pmp csrs belong to you, so this hart_id, I think it is not useful.

baiqiang728 replied:
My intention in reserving this parameter is that perhaps users may use different method to get the number of pmp entries. Such as using constant define in a macro. And the pmp entries of each core may be different.

WiseCrohn reviewed (approved):
I would like to make multiple small fixes to the comments and the English.  But it is easier to submit this PR now and I will submit my own PR for these fixes (with no change to the API).  Following that, I may have further suggestions for extending the API to use a shadow memory area.  We can discuss that separately.
